Blood Cell Composition
========================================================

# Correlate sample methylation signals with each blood type
Gran:neutrophils (granulocytes)
CD8T: CD8+ T cells 
CD4T: CD4+ T cells
NK: CD56+ natural killer cells 
Bcell: CD19+ B cells

## Libraries
```{r}
setwd("~/Documents/Blood_Brain/")
library(reshape)
library(ggplot2)
library(RColorBrewer)
library(mclust)
```

## Load combatted data with no replicates
```{r}
load("BLBR_Beta_Combatted_Mval_bmiqonly.RData") #441198     70
```

#### Any meta data
```{r}
meta<-read.csv("~/Documents/Blood_Brain/SampleInfo.csv")
fix_names<-unlist(lapply(meta$X, function(x) gsub(" ","", x , fixed=TRUE)))
meta$X<-fix_names
meta<-meta[which(meta$X%in%colnames(combat_BLBR_Beta)),]
meta<-meta[match(colnames(combat_BLBR_Beta), meta$X),]
```

### Actualy cell composition with condifential hack
```{r}
library(FlowSorted.Blood.450k)
library(IlluminaHumanMethylation450kmanifest)
library(quadprog)

RGset = as.data.frame(combat_BLBR_Beta)

source("R_Scripts/ECC2.R")

cellprop2 <- ECC2(RGset)
cellprop<-as.data.frame(cellprop2$counts)
cellprop$Sample<-rownames(cellprop)

save(cellprop, file="BLBR_cell_proportions_Mcombat_bmiqonly.RData")




# PLot
PBMC_meta<-subset(meta, TissueType=="PBMC")
cellprop_PBMC<-cellprop[which(rownames(cellprop)%in%PBMC_meta$X),]
cellprop_melt<-melt(cellprop_PBMC, id="Sample")
ggplot(cellprop_melt, aes(variable, value))+
  geom_point(shape=19, position = position_jitter(w = 0.1))+theme_bw()+scale_color_manual(values=c("grey","red"))+xlab("Sample")+ylab("Cell Type Proportion")
```




## Megans code to correct for cell type
```{r}
Blood_beta<-combat_BLBR_Beta[,which(colnames(combat_BLBR_Beta)%in%PBMC_meta$X)]

# Impute NAs
imputeMedianv3<-function(x) apply(x, 1, function(x){x[is.na(x)]<-median(x, na.rm=T); x}) #impute with row mean
Blood_beta_imputed<-t(imputeMedianv3(Blood_beta))


avebeta.lm<-apply(Blood_beta_imputed, 1, function(x){
  cellprop_PBMC[colnames(Blood_beta_imputed),]->blood
  lm(x~CD8T+CD4T+NK+Bcell+Mono,data=blood)
})
residuals<-t(sapply(avebeta.lm, function(x)residuals(summary(x))))
colnames(residuals)<-colnames(Blood_beta_imputed)
adj.residuals<-residuals+matrix(apply(Blood_beta_imputed, 1, mean), nrow=nrow(residuals), ncol=ncol(residuals))
save(adj.residuals, file="adjusted_betas_blood_Mcombat_bmiqonly.RData")

## Check how adjusted for cell type
RGset = as.data.frame(adj.residuals)
source("R_Scripts/ECC2.R")

cellprop2 <- ECC2(RGset)
cellprop_adjusted<-as.data.frame(cellprop2$counts)
cellprop_adjusted$Sample<-rownames(cellprop_adjusted)
save(cellprop_adjusted, file="cellprop_adjusted_Mcombat_bmiqonly.RData")
```

```{r}
##Compare PLot
load("BLBR_cell_proportions_Mcombat_bmiqonly.RData")
load("cellprop_adjusted_Mcombat_bmiqonly.RData")

PBMC_meta<-subset(meta, TissueType=="PBMC")

cellprop_PBMC<-cellprop[which(rownames(cellprop)%in%PBMC_meta$SampleID),]
cellprop_melt<-melt(cellprop_PBMC, id="Sample")
cellprop_melt$Data<-"unadjusted"

cellprop_PBMC_adjusted<-cellprop_adjusted[which(rownames(cellprop_adjusted)%in%PBMC_meta$SampleID),]
cellprop_melt_adjusted<-melt(cellprop_PBMC_adjusted, id="Sample")
cellprop_melt_adjusted$Data<-"adjusted"

cellprop_melt<-rbind(cellprop_melt, cellprop_melt_adjusted)

ggplot(cellprop_melt, aes(variable, value))+
  geom_point(shape=19, position = position_jitter(w = 0.1))+theme_bw()+scale_color_manual(values=c("grey","red"))+xlab("Sample")+ylab("Cell Type Proportion")+facet_wrap(~Data)

```











#### CETS
HTT_Leavitt CETS Brain Cell Type Correction
========================================================
## By: Sumaiya Islam
## Date: September 29, 2014

### Script contents:
- CETS brain cell type correction for normalized, combat-corrected cortex-only dataset (n = 13)
- CETS brain cell type correction for cortex samples (n = 5) of normalized matched dataset
- CETS brain cell type correction for normalized entire dataset (n = 18)

##### We will use an open-source R package, Cell EpigenoType Specific (CETS) mapper, designed for the quantification and normalization of differing neuronal proportions in genome-scale DNA methylation datasets. The application of CETS quantification and transformation can reduce cell type heterogeneity and improve replicability of epigenetic findings in the brain across cohorts. It is important to correct for differences in cell type proportions in methylation data as these differences can lead to false discoveries in the methylation analysis (Guintivano et al 2013 *Epigenetics*).

We will initially set our working directory and load our libraries.

```{r}
install.packages("cets_0.99.2.tar.gz",repos = NULL, type="source")
```


```{r}
library(cets)
library(wateRmelon)
library(lumi)
```

```{r}
# load cortex only dataset (after normalization and ComBat)
load("BLBR_Beta_Combatted_Mval_bmiqonly.RData") #460199     70
meta<-read.csv("~/Documents/Blood_Brain/SampleInfo.csv")
fix_names<-unlist(lapply(meta$X, function(x) gsub(" ","", x , fixed=TRUE)))
meta$X<-fix_names
meta<-meta[which(meta$X%in%colnames(combat_BLBR_Beta)),]
meta<-meta[match(colnames(combat_BLBR_Beta), meta$X),]

BLBR_brain_betas<-combat_BLBR_Beta[,which(meta$TissueType!="PBMC")]
meta_brain<-meta[which(meta$X%in%colnames(BLBR_brain_betas)),]
meta_brain<-meta_brain[match(colnames(BLBR_brain_betas), meta_brain$X),]
```


Load calibration data set

The calibration dataset included in the cetsBrain package contains FACS-sorted neuron and glia samples from a cohort of depression patients and control subjects. The package contains two objects: A data matrix (brain) of sample methylation profiles for the top 10,000 most differentially methylated probes between neurons and glia and (pdBrain) a sample annotation data frame:

```{r}
# load "brain dataset" from data file in cetsBrain
load("cetsBrain/data/cetsBrain.rda") # click on cetsBrain.rda file to place in workspace
```


Create the neuron and glia reference profiles:
```{r}
modelIdx <- list(neuron = pdBrain$celltype == "N", glia = pdBrain$celltype ==  "G")
 # getReference returns a 2-column matrix, representing reference profiles for the two cell types.
refProfile <- getReference(brain, modelIdx)
```

Estimate the neuronal proportion:

The estProportion function returns an estimate of the percentage of cell type in the first column of its profile argument (neurons in this case). 
```{r}
prop <- estProportion(BLBR_brain_betas, profile = refProfile)
prop<-as.data.frame(prop)
prop$glia<-apply(prop,1,function(x) 1-x)
colnames(prop)<- c("neuron", "glia")

save(prop, file="Brain_cell_composition_Mcombat_bmiqonly.RData")

# Plot
prop$sample<-rownames(prop)
porp_melt<-melt(prop)
prop_merge<-merge(porp_melt, meta_brain, by.x="sample", by.y="X")

ggplot(prop_merge, aes(variable, value, color=TissueType))+
  geom_point(shape=19, position=position_jitter(width=0.15))+
  theme_bw()+
  scale_color_manual(values=c("#fb6a4a","#ef3b2c","#cb181d"))
```


We will now perform cell-type correction based on the neuronal/glial proportions 

```{r}
# fit methylation data for each probe in the dataset by the neuronal proportion

# Impute NAs
imputeMedianv3<-function(x) apply(x, 1, function(x){x[is.na(x)]<-median(x, na.rm=T); x}) #impute with row mean
BLBR_brain_betas_imputed<-t(imputeMedianv3(BLBR_brain_betas))

# Split by Brain Region
BR7_betas_imputed<-BLBR_brain_betas_imputed[,which(colnames(BLBR_brain_betas_imputed)%in%
                                                    meta_brain$X[which(meta_brain$TissueType=="BRAIN7")])]
BR10_betas_imputed<-BLBR_brain_betas_imputed[,which(colnames(BLBR_brain_betas_imputed)%in%
                                                    meta_brain$X[which(meta_brain$TissueType=="BRAIN10")])]
BR20_betas_imputed<-BLBR_brain_betas_imputed[,which(colnames(BLBR_brain_betas_imputed)%in%
                                                    meta_brain$X[which(meta_brain$TissueType=="BRAIN20")])]

#Adjust Betas for cell types
Adjust<-function(BR){
avebeta.lm<-apply(BR, 1, function(x){ ## FULL Adjustment!
    brain.sub<-prop[colnames(BR),]
    lm(x~neuron,data=brain.sub)
  })
  residuals<-t(sapply(avebeta.lm, function(x)residuals(summary(x))))
  colnames(residuals)<-colnames(BR)
  adj.residuals<-residuals+matrix(apply(BR, 1, mean), nrow=nrow(residuals), ncol=ncol(residuals))
adj.residuals}

BR7_adjusted<-Adjust(BR7_betas_imputed)
BR10_adjusted<-Adjust(BR10_betas_imputed)
BR20_adjusted<-Adjust(BR20_betas_imputed)

adj.residuals<-cbind(BR7_adjusted, BR10_adjusted, BR20_adjusted)

adj.residuals<-adj.residuals[,match(meta_brain$X, colnames(adj.residuals))]

save(adj.residuals, file="adjusted_betas_brain_Mcombat_bmiqonly_sepBrainregion.RData")




# do composition again (on only complete cases, since NAs introduced by adjustment)
prop_adjusted<- estProportion(adj.residuals, profile = refProfile)
prop_adjusted<-as.data.frame(prop_adjusted)
prop_adjusted$glia<-apply(prop_adjusted,1,function(x) 1-x)
colnames(prop_adjusted)<- c("neuron", "glia")
save(prop_adjusted, file="cellprop_brain_adjusted_Mcombat_bmiqonly_sepBrainregion.RData") 
```

```{r}
##Compare PLot
load("Brain_cell_composition_Mcombat_bmiqonly.RData")
load("cellprop_brain_adjusted_Mcombat_bmiqonly.RData")
prop_adjusted_alltogether<-prop_adjusted
load("cellprop_brain_adjusted_Mcombat_bmiqonly_sepBrainregion.RData")

meta_brain<-subset(meta, TissueType!="PBMC")

cellprop_brain<-prop[which(rownames(prop)%in%meta_brain$SampleID),]
cellprop_brain$Sample<-rownames(cellprop_brain)
cellprop_melt<-melt(cellprop_brain, id="Sample")
cellprop_melt$Data<-"unadjusted"

cellprop_brain_adjusted_alltogether<-prop_adjusted_alltogether[which(rownames(prop_adjusted_alltogether)%in%meta_brain$SampleID),]
cellprop_brain_adjusted_alltogether$Sample<-rownames(cellprop_brain_adjusted_alltogether)
cellprop_melt_adjusted_alltogether<-melt(cellprop_brain_adjusted_alltogether, id="Sample")
cellprop_melt_adjusted_alltogether$Data<-"adjusted_alltogether"

cellprop_brain_adjusted<-prop_adjusted[which(rownames(prop_adjusted)%in%meta_brain$SampleID),]
cellprop_brain_adjusted$Sample<-rownames(cellprop_brain_adjusted)
cellprop_melt_adjusted<-melt(cellprop_brain_adjusted, id="Sample")
cellprop_melt_adjusted$Data<-"adjusted_seperated_regions"

cellprop_melt_brain<-rbind(cellprop_melt, cellprop_melt_adjusted,cellprop_melt_adjusted_alltogether)

ggplot(cellprop_melt_brain, aes(variable, value))+
  geom_point(shape=19, position = position_jitter(w = 0.1))+theme_bw()+scale_color_manual(values=c("grey","red"))+xlab("Sample")+ylab("Cell Type Proportion")+facet_wrap(~Data)

# Split by brian type
cellprop_melt_braintype<-merge(cellprop_melt_brain, meta, by.x="Sample", by.y="SampleID")

cellprop_melt_braintype$Data <- factor(cellprop_melt_braintype$Data,c("unadjusted", "adjusted_alltogether","adjusted_seperated_regions"))

#with all together
ggplot(cellprop_melt_braintype, aes(variable, value))+
  geom_boxplot(color="grey", outlier.shape=NA)+
  geom_point(aes(color=as.factor(variable)), shape=19, position = position_jitter(w = 0.2))+theme_bw()+xlab("Sample")+ylab("Cell Type Proportion")+
  facet_wrap(Data~TissueType, scales="free_x")+scale_color_manual(values=c("#4575b4", "#74add1"),guide=FALSE)

# just seperated
ggplot(subset(cellprop_melt_braintype, Data!="adjusted_alltogether"), aes(variable, value))+
  geom_boxplot(color="grey", outlier.shape=NA)+
  geom_point(aes(color=as.factor(variable)), shape=19, position = position_jitter(w = 0.2))+theme_bw()+xlab("Sample")+ylab("Cell Type Proportion")+
  facet_wrap(Data~TissueType, scales="free_x")+scale_color_manual(values=c("#4575b4", "#74add1"),guide=FALSE)


```



##Compare PLot ALL
```{r}
#Brain
load("Brain_cell_composition_Mcombat_bmiqonly.RData")
load("cellprop_brain_adjusted_Mcombat_bmiqonly.RData")
meta_brain<-subset(meta, TissueType!="PBMC")
cellprop_brain<-prop[which(rownames(prop)%in%meta_brain$SampleID),]
cellprop_brain$Sample<-rownames(cellprop_brain)
cellprop_melt<-melt(cellprop_brain, id="Sample")
cellprop_melt$Data<-"unadjusted"
cellprop_brain_adjusted<-prop_adjusted[which(rownames(prop_adjusted)%in%meta_brain$SampleID),]
cellprop_brain_adjusted$Sample<-rownames(cellprop_brain_adjusted)
cellprop_melt_adjusted<-melt(cellprop_brain_adjusted, id="Sample")
cellprop_melt_adjusted$Data<-"adjusted"
cellprop_melt_brain<-rbind(cellprop_melt, cellprop_melt_adjusted)
cellprop_melt_brain$Tissue<-"Brain"

#Blood
load("BLBR_cell_proportions_Mcombat_bmiqonly.RData")
load("cellprop_adjusted_Mcombat_bmiqonly.RData")
PBMC_meta<-subset(meta, TissueType=="PBMC")
cellprop_PBMC<-cellprop[which(rownames(cellprop)%in%PBMC_meta$SampleID),]
cellprop_melt<-melt(cellprop_PBMC, id="Sample")
cellprop_melt$Data<-"unadjusted"
cellprop_PBMC_adjusted<-cellprop_adjusted[which(rownames(cellprop_adjusted)%in%PBMC_meta$SampleID),]
cellprop_melt_adjusted<-melt(cellprop_PBMC_adjusted, id="Sample")
cellprop_melt_adjusted$Data<-"adjusted"
cellprop_melt_blood<-rbind(cellprop_melt, cellprop_melt_adjusted)
cellprop_melt_blood$Tissue<-"Blood"
cellprop_melt<-rbind(cellprop_melt_blood, cellprop_melt_brain)
cellprop_melt<-merge(cellprop_melt, meta[,c(2,3)], by.y="SampleID", by.x="Sample")


# plot
cellprop_melt$Data <- as.factor(cellprop_melt$Data)
cellprop_melt$Data <- factor(cellprop_melt$Data,c("unadjusted", "adjusted"))

ggplot(cellprop_melt, aes(variable, value, color=as.factor(variable)))+
  geom_point(shape=19, position = position_jitter(w = 0.2))+theme_bw()+xlab("Sample")+ylab("Cell Type Proportion")+
  facet_wrap(Tissue~Data, scales="free_x")+scale_color_manual(values=c("#a50026", "#bd0026", "#e31a1c", "#fc4e2a", "#fd8d3c", "#feb24c", "#4575b4", "#74add1"),guide=FALSE)

## Cell Type Correlation
Proportions_all<-subset(cellprop_melt, Data=="unadjusted")[,c(2,3,6)]
Proportions_all<-cast(Proportions_all, SubjectNumC~variable, mean)
lapply(2:ncol(Proportions_all), function(x) sapply(2:ncol(Proportions_all), function(y) {
  cor.test(Proportions_all[,x], Proportions_all[,y], alternative = "two.sided", method = "pearson", na.action = na.omit)$p.value}))
correlations<-as.data.frame(cor(as.matrix(Proportions_all[2:9]), use="complete.obs"))
colnames(correlations)<-colnames(Proportions_all[2:9])
correlations$Cell_Type<-as.factor(colnames(Proportions_all[2:9]))
correlation_melt<-melt(correlations)
correlation_melt$Cell_Type <- factor(correlation_melt$Cell_Type,levels(correlation_melt$variable))

#heat map
ggplot(correlation_melt, aes(Cell_Type,variable, fill = value)) +
  geom_tile(color = "black",size=0.5) +
  theme_gray(8)+theme(axis.text = element_text(size =10, color="black"),
          axis.title = element_text(size =15),
          axis.text.x = element_text(angle = 90, hjust = 1))+
  scale_fill_gradientn(colours = cols, 
                         values = rescale(c(-0.65, 0,0.65)),
                         guide = "colorbar", limits=c(-1, 1))+
  xlab("Cell Type")+ylab("Cell Type")
```





### REMOVE 10000 CETS probes
### reomve X Blood probes




## Compare replicates
```{r}
meta<-read.csv("~/Documents/Blood_Brain/SampleInfo.csv")
fix_names<-unlist(lapply(meta$X, function(x) gsub(" ","", x , fixed=TRUE)))
meta$X<-fix_names


load("adjusted_betas_blood_Mcombat_bmiqonly.RData")
adj.residuals_blood<-adj.residuals
load("adjusted_betas_brain_Mcombat_bmiqonly_sepBrainregion.RData")
adj.residuals_brain<-adj.residuals

adj.residuals_blood_brain<-cbind(adj.residuals_blood, adj.residuals_brain)


adj.residuals_blood_brain<-adj.residuals_blood_brain[,match(meta$X, colnames(adj.residuals_blood_brain))]
save(adj.residuals_blood_brain, file="BLBR_Beta_Combatted_Mval_bmiqonly_CellCorrected_brSep.RData") # Includes replicates




# Compare Pre and post normalization and combat REPLICATES and cell adjust
```{r}
load("BLBR_methlylumi_second.RData")#original
load("BLBR_norm_filtered.RData") # BMIQ
load("BLBR_Beta_Combatted_Mval_bmiqonly.RData") # Includes replicates
load("BLBR_Beta_Combatted_Mval_bmiqonly_CellCorrected_brSep.RData") # Includes replicates

beta_BLBR.2<-betas(BLBR.2)
beta_BLBR.2_bmiq<-BLBR_Beta

replicates<-data.frame(Sample_ID=meta$SampleID[which(meta$Replicate=="Y")],
                       X=meta$X[which(meta$Replicate=="Y")])
replicates$Pair<-c("PBMC169","BA7250","PBMC250","BA7250","BA10248","PBMC169","BA10248","PBMC250")

original_rep<-beta_BLBR.2[,which(colnames(beta_BLBR.2)%in%replicates$Sample_ID)]
bmiq_rep<-beta_BLBR.2_bmiq[,which(colnames(beta_BLBR.2_bmiq)%in%replicates$X)]
combat_rep<-combat_BLBR_Beta[,which(colnames(combat_BLBR_Beta)%in%replicates$X)]
cell_adjusted_rep<-adj.residuals_blood_brain[,which(colnames(adj.residuals_blood_brain)%in%replicates$X)]


Data_cors<-function(df) {data.frame(Comparison=c("PBMC169", "BA7250",  "PBMC250", "BA10248"), 
                                    Correlation=c(cor(df[,1],df[,6], use="complete.obs"),
                                                  cor(df[,2],df[,4], use="complete.obs"),
                                                  cor(df[,3],df[,8], use="complete.obs"),
                                                  cor(df[,5],df[,7], use="complete.obs")))}

original_rep_cor<-Data_cors(original_rep)
bmiq_rep_cor<-Data_cors(bmiq_rep)
combat_rep_cor<-Data_cors(combat_rep)
cell_adjusted_rep_cor<-Data_cors(cell_adjusted_rep)

original_rep_cor$Data<-"Original"
bmiq_rep_cor$Data<-"BMIQ Norm"
combat_rep_cor$Data<-"BMIQ Norm Combat"
cell_adjusted_rep_cor$Data<-"BMIQ Norm \n Combat Cell Adjust"

rep_cor<-rbind(original_rep_cor,bmiq_rep_cor,combat_rep_cor,cell_adjusted_rep_cor)
rep_cor$Data <- factor(rep_cor$Data, levels=c("Original","BMIQ Norm", "BMIQ Norm Combat", "BMIQ Norm \n Combat Cell Adjust"))
ggplot(rep_cor, aes(Data, Correlation, group=Comparison, color=Comparison))+geom_line(size=2)+theme_bw()+ylab("Replicate Correlation")
```

## Remover replicates
```{r}
rm_replicates<-c("BA7250rep", "PBMC169rep", "BA102482", "PBMC250rep")

#cell comp adjusted
combat_BLBR_Beta<-adj.residuals_blood_brain
combat_BLBR_Beta<-combat_BLBR_Beta[,which(!(colnames(combat_BLBR_Beta)%in%rm_replicates))]# 441198 n=66
save(combat_BLBR_Beta, file="Second_cell_adjusted_Mcombat_Bmiq_BLBR_Beta_noreplicates.RData") #441198     66
# unadjusted
load("BLBR_Beta_Combatted_Mval_bmiqonly.RData") #441198     70
combat_BLBR_Beta<-combat_BLBR_Beta[,which(!(colnames(combat_BLBR_Beta)%in%rm_replicates))]# 441198 n=66
save(combat_BLBR_Beta, file="cell_unadjusted_Mcombat_Bmiq_BLBR_Beta_noreplicates.RData") #441198     66
```



## How many CpGs are dramatically changed by cell type correction
```{r}
load("cell_adjusted_Mcombat_Bmiq_BLBR_Beta_noreplicates.RData") #441198     66
combat_BLBR_Beta_adjusted<-combat_BLBR_Beta
load("cell_unadjusted_Mcombat_Bmiq_BLBR_Beta_noreplicates.RData") #441198     66
combat_BLBR_Beta_unadjusted<-combat_BLBR_Beta

Delta_beta<-abs(as.matrix(combat_BLBR_Beta_adjusted)-as.matrix(combat_BLBR_Beta_unadjusted))
Delta_beta_means<-rowMeans(abs(Delta_beta))

PBMC_meta<-meta[which(meta$TissueType=="PBMC"),]
BRAIN7_meta<-meta[which(meta$TissueType=="BRAIN7"),]
BRAIN10_meta<-meta[which(meta$TissueType=="BRAIN10"),]
BRAIN20_meta<-meta[which(meta$TissueType=="BRAIN20"),]

Blood_Delta_beta<-Delta_beta[,which(colnames(Delta_beta)%in%PBMC_meta$X)]
BR7_Delta_beta<-Delta_beta[,which(colnames(Delta_beta)%in%BRAIN7_meta$X)]
BR10_Delta_beta<-Delta_beta[,which(colnames(Delta_beta)%in%BRAIN10_meta$X)]
BR20_Delta_beta<-Delta_beta[,which(colnames(Delta_beta)%in%BRAIN20_meta$X)]

Blood_Delta_beta_means<-rowMeans(abs(Blood_Delta_beta))
length(Blood_Delta_beta_means[which(Blood_Delta_beta_means>0.1)]) 

BR7_Delta_beta_means<-rowMeans(abs(BR7_Delta_beta))
length(BR7_Delta_beta_means[which(BR7_Delta_beta_means>0.1)])

BR10_Delta_beta_means<-rowMeans(abs(BR10_Delta_beta))
length(BR10_Delta_beta_means[which(BR10_Delta_beta_means>0.1)]) #most varibility of brain regions and most adjusted CpGs

BR20_Delta_beta_means<-rowMeans(abs(BR20_Delta_beta))
length(BR20_Delta_beta_means[which(BR20_Delta_beta_means>0.1)])


```





# Differential methylation between Cell adjusted and unadjusted 
#Not worth doing because porbes will be adjusted in different directions and therefore nothing is significant
```{r}

PBMC_meta<-meta[which(meta$TissueType=="PBMC"),]
BRAIN7_meta<-meta[which(meta$TissueType=="BRAIN7"),]
BRAIN10_meta<-meta[which(meta$TissueType=="BRAIN10"),]
BRAIN20_meta<-meta[which(meta$TissueType=="BRAIN20"),]

#adjusted
Blood_betaad<-combat_BLBR_Beta_adjusted[,which(colnames(combat_BLBR_Beta_adjusted)%in%PBMC_meta$X)]
BR7_betaad<-combat_BLBR_Beta_adjusted[,which(colnames(combat_BLBR_Beta_adjusted)%in%BRAIN7_meta$X)]
BR10_betaad<-combat_BLBR_Beta_adjusted[,which(colnames(combat_BLBR_Beta_adjusted)%in%BRAIN10_meta$X)]
BR20_betaad<-combat_BLBR_Beta_adjusted[,which(colnames(combat_BLBR_Beta_adjusted)%in%BRAIN20_meta$X)]

#unadjusted
Blood_beta<-combat_BLBR_Beta_unadjusted[,which(colnames(combat_BLBR_Beta_unadjusted)%in%PBMC_meta$X)]
BR7_beta<-combat_BLBR_Beta_unadjusted[,which(colnames(combat_BLBR_Beta_unadjusted)%in%BRAIN7_meta$X)]
BR10_beta<-combat_BLBR_Beta_unadjusted[,which(colnames(combat_BLBR_Beta_unadjusted)%in%BRAIN10_meta$X)]
BR20_beta<-combat_BLBR_Beta_unadjusted[,which(colnames(combat_BLBR_Beta_unadjusted)%in%BRAIN20_meta$X)]


#Combine and make new meta
Blood_beta<-cbind(Blood_beta, Blood_betaad)
BR7_beta<-cbind(BR7_beta, BR7_betaad)
BR10_beta<-cbind(BR10_beta, BR10_betaad)
BR20_beta<-cbind(BR20_beta, BR20_betaad)

BloodCorrectedornot<-c(rep("Unadjusted", 16), rep("Adjusted", 16))
BR7Correctedornot<-c(rep("Unadjusted", 17), rep("Adjusted", 17))
BR10Correctedornot<-c(rep("Unadjusted", 17), rep("Adjusted", 17))
BR20Correctedornot<-c(rep("Unadjusted", 16), rep("Adjusted", 16))

save(Blood_beta, BR7_beta, BR10_beta, BR20_beta,
     BloodCorrectedornot,BR7Correctedornot, BR10Correctedornot,BR20Correctedornot,
     file="Differential_meth_cellcomp.RData")


#Pval 
p<-sapply(1:nrow(hmC_Beta_ordered), function(x) {z<-lm(unlist(hmC_Beta_ordered[x,]) ~ Sentrix_ordered$BS)
                                                coef(summary(z))["Sentrix_ordered$BSoxBS", "Pr(>|t|)"]})
Multi_test_corr_relaxed<-p.adjust(p, method = "fdr", n = length(p))
stat_hits<-combat_br10_bld[which(Multi_test_corr_relaxed<=0.001),] #asymptotic Pvalue 153567 probes
#Delta beta
BS<-rowMeans(combat_br10_bld[,which(Sentrix_ordered$BS=="BS")])
oxBS<-rowMeans(combat_br10_bld[,which(Sentrix_ordered$BS=="oxBS")])
delbeta<-BS-oxBS
bio_hits<-hmC_Beta_ordered[which(abs(delbeta)>=0.1),] #asymptotic Pvalue 25917 probes


save(p, Multi_test_corr_relaxed, stat_hits, delbeta, bio_hits, file="~/Documents/Side/Hydroxy_Methyl/Differemtial_methylation_hydroxy.RData")
```


## Grant plot
```{r}


Blood_Delta_beta_means<-rowMeans(abs(Blood_Delta_beta))
length(Blood_Delta_beta_means[which(Blood_Delta_beta_means>0.1)]) 

BR7_Delta_beta_means<-rowMeans(abs(BR7_Delta_beta))
length(BR7_Delta_beta_means[which(BR7_Delta_beta_means>0.1)])

BR10_Delta_beta_means<-rowMeans(abs(BR10_Delta_beta))
length(BR10_Delta_beta_means[which(BR10_Delta_beta_means>0.1)]) #most varibility of brain regions and most adjusted CpGs

BR20_Delta_beta_means<-rowMeans(abs(BR20_Delta_beta))
length(BR20_Delta_beta_means[which(BR20_Delta_beta_means>0.1)])

cellprop_melt_sd<-merge(cellprop_melt, meta, by.x="Sample", by.y="X")

tapply(cellprop_melt_sd$value, list(cellprop_melt_sd$TissueType, cellprop_melt_sd$variable), sd)

x<-data.frame(Adjusted_CpGs=c(72, 4229, 117), sd=rowMeans(tapply(cellprop_melt_sd$value, list(cellprop_melt_sd$TissueType, cellprop_melt_sd$variable), sd), na.rm=T)[1:3])

ggplot(x, aes(sd, Adjusted_CpGs))+geom_line()


load("Brain_cell_composition_Mcombat_bmiqonly.RData")
load("cellprop_brain_adjusted_Mcombat_bmiqonly.RData")
meta_brain<-subset(meta, TissueType!="PBMC")
cellprop_brain<-prop[which(rownames(prop)%in%meta_brain$SampleID),]
cellprop_brain$Sample<-rownames(cellprop_brain)
cellprop_melt<-melt(cellprop_brain, id="Sample")
cellprop_melt$Data<-"unadjusted"
cellprop_brain_adjusted<-prop_adjusted[which(rownames(prop_adjusted)%in%meta_brain$SampleID),]
cellprop_brain_adjusted$Sample<-rownames(cellprop_brain_adjusted)
cellprop_melt_adjusted<-melt(cellprop_brain_adjusted, id="Sample")
cellprop_melt_adjusted$Data<-"adjusted"
cellprop_melt_brain<-rbind(cellprop_melt, cellprop_melt_adjusted)
cellprop_melt_brain$Tissue<-"Brain"

#Blood
load("BLBR_cell_proportions_Mcombat_bmiqonly.RData")
load("cellprop_adjusted_Mcombat_bmiqonly.RData")
PBMC_meta<-subset(meta, TissueType=="PBMC")
cellprop_PBMC<-cellprop[which(rownames(cellprop)%in%PBMC_meta$SampleID),]
cellprop_melt<-melt(cellprop_PBMC, id="Sample")
cellprop_melt$Data<-"unadjusted"
cellprop_PBMC_adjusted<-cellprop_adjusted[which(rownames(cellprop_adjusted)%in%PBMC_meta$SampleID),]
cellprop_melt_adjusted<-melt(cellprop_PBMC_adjusted, id="Sample")
cellprop_melt_adjusted$Data<-"adjusted"
cellprop_melt_blood<-rbind(cellprop_melt, cellprop_melt_adjusted)
cellprop_melt_blood$Tissue<-"Blood"
cellprop_melt<-rbind(cellprop_melt_blood, cellprop_melt_brain)
cellprop_melt<-merge(cellprop_melt, meta[,c(2,3)], by.y="SampleID", by.x="Sample")


# plot
cellprop_melt$Data <- as.factor(cellprop_melt$Data)
cellprop_melt$Data <- factor(cellprop_melt$Data,c("unadjusted", "adjusted"))

cellprop_melt<-merge(cellprop_melt, meta, by.x="Sample", by.y="X")
levels(cellprop_melt$TissueType)<-c("Brodmann brain area 10",
                                    "Brodmann brain area 20",
                                    "Brodmann brain area 7",
                                    "Blood")


ggplot(subset(cellprop_melt, Data=="unadjusted"), aes(variable, value, color=as.factor(variable)))+
  geom_boxplot(color="grey", outlier.size=NA)+
  geom_point(shape=19, position = position_jitter(w = 0.2))+theme_bw()+xlab("Sample")+ylab("Cell Type Proportion")+
  facet_wrap(~TissueType, scales="free_x")+scale_color_manual(values=c("#a50026", "#bd0026", "#e31a1c", "#fc4e2a", "#fd8d3c", "#feb24c", "#4575b4", "#74add1"),guide=FALSE)



```

